### ISSUE TITLE
Add "fork" method to DenoPermissions

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno#4779]
To have `Deno` namespace in the worker, we'll want to provide a map of permissions worker should have. Worker's set of permissions mustn't escalate permissions of parent worker/Deno process. 

To achieve that a method called `fork` should be added to `DenoPermissions` struct. `fork` would take as many parameters as there are fields on `DenoPermissions` and return new `DenoPermissions` instance after asserting that none of the permissions passed as arguments escalate parent permissions. If one tried to fork permissions with escalation then `OpError::permission_denied` should be returned with appropriate message indicating which permissions are being escalated.

Pseudo code:
```rust
impl DenoPermissions {
  fn fork(&self, allow_net: bool, net_whitelist: Vec<String>, ...) -> Result<DenoPermissions, OpError>;
}
``` 

This is an easy issue, only `*_whitelist` processing might be a bit complicated.

**Comment by bartlomieju** [comment]
Prerequisite for #3998

**Comment by humancalico** [comment]
@bartlomieju I'll work on this.

**Comment by jafern14** [comment]
@bartlomieju is this task still necessary? #3998 is closed. trying to figure out the value of this given that closed dep

**Comment by bartlomieju** [comment]
@jefern14 yes it's needed. Currently in worker permissions are inherited from process and we want to make them configurable. See #4867

**Comment by jafern14** [comment]
great thanks for the update. I'll be looking into this as well after work hours.

**Comment by jafern14** [comment]
@bartlomieju do you mind explaining a bit more of what you mean when you say "permissions escalation"? The way I'm looking at this having almost no knowledge of how this works is that a forked worker would inherit all the same permissions of it's parent. Your description makes me doubt my assumption so I'd like to understand prior implementing anything.

**Comment by bartlomieju** [comment]
> @bartlomieju do you mind explaining a bit more of what you mean when you say "permissions escalation"? The way I'm looking at this having almost no knowledge of how this works is that a forked worker would inherit all the same permissions of it's parent. Your description makes me doubt my assumption so I'd like to understand prior implementing anything.

That's the case right now, when you create worker with `Deno` namespace it inherits parent process permissions. We want to make this option configurable so you can explicitly specify permissions for worker (eg. `{ read: ["/tmp"], net: true }`). Escalation means that requested permissions for worker cannot be more permissive than main process permissions - ie. if you run program with only `--allow-read` flag the workers created cannot request that `net` and `write` are available to them.

**Comment by jafern14** [comment]
> ie. if you run program with only --allow-read flag the workers created cannot request that net and write are available to them.

since the `--allow-read` flag looks like it's explicitly calling out read permissions does that mean if the parent process has either `net` and/or `write` then the forked process would still also get those permissions? or are you saying that `--allow-read` is explicit and only allows that permissions? If that's the case then I'd imagine that any `--allow-*` flag is also explicit by nature.

**Comment by Axighi** [comment]
@bartlomieju  I'm trying to work on this.

**Comment by humancalico** [comment]
@Axighi I already have an open PR. Although I do have to update it since `DenoPermissions` got renamed to `Permissions`

**Comment by ktfth** [comment]
The fork process just can work on Workers or have other scenarios?

**Comment by fabrizzioalco** [comment]
Hey, is this done? I want to contribute to this project

**Comment by bartlomieju** [comment]
Hi @fabrizzioalco, yes this issue was resolved in #5693. Try looking for issues tagged with "good first issue".
